src/uwtools/rocoto.py

```diff
@@ -25,20 +35,110 @@ def _add_jobname(tree: dict) -> None:
-    :param rendered_output: Path to write rendered XML file.
     """
     values = YAMLConfig(input_yaml)
-    tasks = values["tasks"]
+    tasks = values["workflow"]["tasks"]
     if isinstance(tasks, dict):
         _add_jobname(tasks)
 
+
+def _rocoto_schema() -> str:
+    """
+    The path to the file containing the schema to validate the XML file against.
+    """
+    with resources.as_file(resources.files("uwtools.resources")) as path:
+        return (path / "rocoto.jsonschema").as_posix()
+
+
+def _rocoto_template() -> str:
+    """
+    The path to the file containing the Rocoto workflow document template to render.
+    """
+    with resources.as_file(resources.files("uwtools.resources")) as path:
+        # return (path / "rocoto.jinja2").as_posix()
+        return (path / "schema_with_metatasks.rng").as_posix()
+
+
+# Public functions
+def realize_rocoto_xml(
+    config_file: OptionalPath = None,
+    rendered_output: OptionalPath = None,
+) -> bool:
+    """
+    Realize the given YAML file to XML, using the Rocoto RelaxNG schema. Validate both the YAML and
+    the XML.
+
+    :param input_yaml: Path to YAML input file.
+    :param rendered_output: Path to write rendered XML file.
+    """
+
+    rocoto_schema = _rocoto_schema()
+
+    # Validate the YAML.
+    if uwtools.config.validator.validate_yaml(config_file=config_file, schema_file=rocoto_schema):
+        _add_tasks(config_file)
+        # Render the template to a temporary file.
+        with tempfile.NamedTemporaryFile(delete=False) as temp_file:
+            write_rocoto_xml(
+                config_file=config_file,
+                rendered_output=temp_file.name,
+            )
+            # Validate the XML.
+            if validate_rocoto_xml(input_xml=temp_file.name):
+                # If no issues were detected, save temp file and report success.
+                shutil.move(temp_file.name, str(rendered_output))
+                return True
+        logging.error("Rocoto validation errors identified in %s", temp_file.name)
+        return False
+    logging.error("YAML validation errors identified in %s", config_file)
+    return False
+
+
+def validate_rocoto_xml(input_xml: OptionalPath = None) -> bool:
+    """
+    Given a rendered XML file, validate it against the Rocoto schema.
+
+    :param input_XML: Path to rendered XML file.
+    """
+    rocoto_template = _rocoto_template()
+
+    # Validate the XML.
+    with readable(rocoto_template) as f:
+        schema = etree.RelaxNG(etree.parse(f))
+    tree = etree.parse(input_xml)
+    success = schema.validate(tree)
+
+    # Store validation errors in the main log.
+    errors = str(etree.RelaxNG.error_log).split("\n")
+    log_method = logging.error if len(errors) else logging.info
+    log_method("%s Rocoto validation error%s found", len(errors), "" if len(errors) == 1 else "s")
+    for line in errors:
+        logging.error(line)
+
+    return success
+
+
+def write_rocoto_xml(
```

Comment:
Do we call this function from _outside_ this module, apart from in tests? I don't think so... If not, it should be marked private.

Reply:
Kind of an extension of our earlier discussions about docstrings that said "main entry" or "functions access this", etc. I should bite the bullet and put everything not called `realize()` private

Reply:
I think we have two public functions, which need to be callable from the CLI: `realize_rocoto_xml()` and `validate_rocoto_xml()`. True?